### ISSUE TITLE
Crossfade Method 

### ISSUE THREAD

**Issue by ckeurk** [AzuraCast/AzuraCast#1484]
**Installation method**
Docker

Small problem with Liquidsoap's crossfading method.
If I switch to smart mode, the sequence is very short even if I change the duration of 3 seconds for the overlapping songs.

If I switch to normal mode, the sequence is good, but I get the error message "AzuraCast is Live" still there for me.

The modification to the broadcast titles is the "Custom Cues: Cue-Out Point (seconds)"
Song Length (seconds)
450
Custom Cues: Cue-Out Point (seconds)
400

In other words, the smart mode is like the disabled mode 🥺
_it's the same on the demo version_

**Comment by Vaalyn** [comment]
I played around with the smart crossfade in the demo installation to see if it is working or not. Sometimes it did crossfade sometimes not.

When I looked into the logs this is what I can see when setting the crossfade duration to 10s (which is pretty high):

```
2019/04/27 12:08:18 [smart_cross_6440:3] Analysis: -9.141963dB / -12.665697dB (15.00s / 15.00s)
2019/04/27 12:08:18 [smart_crossfade:3] No transition: using default.
2019/04/27 12:08:33 [lang:3] AzuraCast Feedback Response: OK
2019/04/27 12:08:50 [server:3] New client: 172.18.0.6.
2019/04/27 12:08:50 [server:3] Client 172.18.0.6 disconnected.
2019/04/27 12:08:50 [azuratest_radio_local_1:3] Performing user-requested skip
2019/04/27 12:08:50 [playlist_default:3] Finished with "/var/azuracast/stations/azuratest_radio/media/until_we_meet_again.mp3".
2019/04/27 12:08:50 [decoder:3] Method "MAD" accepted "/var/azuracast/stations/azuratest_radio/media/strolling_through.mp3".
2019/04/27 12:08:50 [azuratest_radio_next_song:3] Prepared "/var/azuracast/stations/azuratest_radio/media/strolling_through.mp3" (RID 6).
2019/04/27 12:08:50 [azuratest_radio_autodj_fallback:3] Switch to azuratest_radio_cue_cut with forgetful transition.
2019/04/27 12:08:51 [smart_cross_6440:3] Analysis: -8.862670dB / -13.406978dB (15.00s / 15.00s)
2019/04/27 12:08:51 [smart_crossfade:3] No transition: using default.
2019/04/27 12:08:55 [server:3] New client: 172.18.0.6.
2019/04/27 12:08:55 [server:3] Client 172.18.0.6 disconnected.
2019/04/27 12:08:55 [azuratest_radio_local_1:3] Performing user-requested skip
2019/04/27 12:08:55 [lang:3] AzuraCast Feedback Response: OK
2019/04/27 12:10:53 [lang:3] AzuraCast Raw Response: annotate:title="Until We Meet Again",artist="Joe Bagale",duration="196.",song_id="51cb8c6254da70288d2f749817403853",media_id="16",genre="16",playlist_id="1":/var/azuracast/stations/azuratest_radio/media/until_we_meet_again.mp3
2019/04/27 12:10:53 [decoder:3] Method "MAD" accepted "/var/azuracast/stations/azuratest_radio/media/until_we_meet_again.mp3".
2019/04/27 12:11:03 [azuratest_radio_next_song:3] Finished with "/var/azuracast/stations/azuratest_radio/media/strolling_through.mp3".
2019/04/27 12:11:03 [azuratest_radio_next_song:3] Prepared "/var/azuracast/stations/azuratest_radio/media/until_we_meet_again.mp3" (RID 8).
2019/04/27 12:11:03 [smart_cross_6440:3] Analysis: -49.837269dB / -12.625719dB (15.03s / 15.05s)
2019/04/27 12:11:03 [smart_crossfade:3] No transition: using default.
2019/04/27 12:11:18 [lang:3] AzuraCast Feedback Response: OK
```

And this is a log when it's set to 6s:
```
2019/04/27 12:14:44 [azuratest_radio_local_1:3] Connection setup was successful.
2019/04/27 12:14:44 [azuratest_radio_local_2:3] Connecting mount /mobile.mp3 for source@127.0.0.1...
2019/04/27 12:14:45 [azuratest_radio_local_2:3] Connection setup was successful.
2019/04/27 12:14:45 [threads:3] Created thread "wallclock_main" (1 total).
2019/04/27 12:14:45 [clock.wallclock_main:3] Streaming loop starts, synchronized with wallclock.
2019/04/27 12:14:45 [azuratest_radio_autodj_fallback:3] Switch to sequence_6423.
2019/04/27 12:14:45 [azuratest_radio_next_song:3] Prepared "/var/azuracast/stations/azuratest_radio/media/c_major_prelude.mp3" (RID 3).
2019/04/27 12:14:45 [azuratest_radio_autodj_fallback:3] Switch to azuratest_radio_cue_cut with forgetful transition.
2019/04/27 12:14:45 [smart_cross_6440:3] Analysis: -infdB / -21.676629dB (5.00s / 5.00s)
2019/04/27 12:14:45 [smart_crossfade:3] new >= old + margin, old <= medium and new <= high.
2019/04/27 12:14:45 [smart_crossfade:3] Do not fade if it's already very low.
2019/04/27 12:14:45 [smart_crossfade:3] Transition: crossed, no fade.
2019/04/27 12:14:45 [lang:3] AzuraCast Feedback Response: OK
2019/04/27 12:16:39 [lang:3] AzuraCast Raw Response: annotate:title="Zen Valley",artist="Josh Kirsch/Media Right Productions",duration="106.",song_id="6d131283c66496c4c8a334b509d34a3a",media_id="4",genre="4",playlist_id="1":/var/azuracast/stations/azuratest_radio/media/zen_valley.mp3
2019/04/27 12:16:39 [decoder:3] Method "MAD" accepted "/var/azuracast/stations/azuratest_radio/media/zen_valley.mp3".
2019/04/27 12:16:49 [azuratest_radio_next_song:3] Finished with "/var/azuracast/stations/azuratest_radio/media/c_major_prelude.mp3".
2019/04/27 12:16:49 [azuratest_radio_next_song:3] Prepared "/var/azuracast/stations/azuratest_radio/media/zen_valley.mp3" (RID 4).
2019/04/27 12:16:49 [smart_cross_6440:3] Analysis: -80.084739dB / -14.234837dB (9.01s / 9.03s)
2019/04/27 12:16:49 [smart_crossfade:3] No transition: using default.
2019/04/27 12:16:58 [lang:3] AzuraCast Feedback Response: OK
2019/04/27 12:18:25 [lang:3] AzuraCast Raw Response: annotate:title="Ticker",artist="Silent Partner",duration="183.",song_id="1d53f2afacfad38dfed92275545a1eaa",media_id="3",genre="3",playlist_id="1":/var/azuracast/stations/azuratest_radio/media/ticker.mp3
2019/04/27 12:18:25 [decoder:3] Method "MAD" accepted "/var/azuracast/stations/azuratest_radio/media/ticker.mp3".
2019/04/27 12:18:35 [azuratest_radio_next_song:3] Finished with "/var/azuracast/stations/azuratest_radio/media/zen_valley.mp3".
2019/04/27 12:18:35 [azuratest_radio_next_song:3] Prepared "/var/azuracast/stations/azuratest_radio/media/ticker.mp3" (RID 5).
2019/04/27 12:18:35 [smart_cross_6440:3] Analysis: -66.480035dB / -18.380318dB (9.02s / 9.06s)
2019/04/27 12:18:35 [smart_crossfade:3] new >= old + margin, old <= medium and new <= high.
2019/04/27 12:18:35 [smart_crossfade:3] Do not fade if it's already very low.
2019/04/27 12:18:35 [smart_crossfade:3] Transition: crossed, no fade.
2019/04/27 12:18:35 [lang:3] AzuraCast Feedback Response: OK
2019/04/27 12:21:19 [lang:3] AzuraCast Raw Response: annotate:title="Hold On a Minute",artist="Silent Partner",duration="146.",song_id="280f593a450342abd06e60a4bf33cd80",media_id="10",genre="10",playlist_id="1":/var/azuracast/stations/azuratest_radio/media/hold_on_a_minute.mp3
2019/04/27 12:21:19 [decoder:3] Method "MAD" accepted "/var/azuracast/stations/azuratest_radio/media/hold_on_a_minute.mp3".
2019/04/27 12:21:29 [azuratest_radio_next_song:3] Finished with "/var/azuracast/stations/azuratest_radio/media/ticker.mp3".
2019/04/27 12:21:29 [azuratest_radio_next_song:3] Prepared "/var/azuracast/stations/azuratest_radio/media/hold_on_a_minute.mp3" (RID 6).
2019/04/27 12:21:30 [smart_cross_6440:3] Analysis: -54.905901dB / -28.670497dB (9.02s / 9.02s)
2019/04/27 12:21:30 [smart_crossfade:3] new >= old + margin, old <= medium and new <= high.
2019/04/27 12:21:30 [smart_crossfade:3] Do not fade if it's already very low.
2019/04/27 12:21:30 [smart_crossfade:3] Transition: crossed, no fade.
2019/04/27 12:21:30 [lang:3] AzuraCast Feedback Response: OK
```

There may be ways to optimize the behaviour of the smart crossfading but in general it will behave in such a way that it calculates the dB volume level of the end of the current track and the start of the next track and then decides:

- If A and B and not too loud and close, fully cross-fade them
- If B is significantly louder than A, only fade-out A
- Do not fade if it's already very low
- Opposite as the previous one.
- Otherwise, A and B are just too loud to overlap nicely, or the difference between them is too large and overlapping would completely mask one of them.

[Here](https://www.liquidsoap.info/doc-1.3.6/smartcrossfade.html) is the page from the liquidsoap docs about smart crossfade, there you can see the script that is used for this function.

I also noticed that there is a new [beta release](https://github.com/savonet/liquidsoap/releases/tag/1.4.0-alpha2) for Liquidsoap for version 1.4.0 (we are currently using 1.3.7 which is the currently stable release) where the `smart_crossfade` function was tweaked a bit and also promoted to be the main `crossfade` function. The general behaviour should still be the same so I'm not sure if that will bring any noticable changes.

@SlvrEagle23 When LS 1.4.0 is released as stable we need to check if we need to adapt anything to the changes in that release. One thing would be changing `smart_crossfade` to `crossfade` in the LS config generation.

**Comment by ckeurk** [comment]
Thank you for your answer. We'll see.

**Comment by SlvrEagle23** [comment]
@Vaalyn We'll definitely need to keep an eye on 1.4.0's release. We'll be able to adopt it immediately on all platforms thanks to our switch to OPAM2, and hopefully it will offer some significant improvements to things like crossfading, which have long been a thorn in our side.
